src/Avalonia.Controls/Control.cs

```diff
@@ -215,18 +279,89 @@ void IVisualBrushInitialize.EnsureInitialized()
         /// <returns>The control that receives the focus adorner.</returns>
         protected virtual IControl? GetTemplateFocusTarget() => this;
 
+        /// <summary>
+        /// Invoked as the first step of marking the control as loaded and raising the
+        /// <see cref="Loaded"/> event.
+        /// </summary>
+        internal void OnLoadedCore()
+        {
+            if (_isLoaded == false)
+            {
+                _isLoaded = true;
+                OnLoaded();
+            }
+        }
+
+        /// <summary>
+        /// Invoked as the first step of marking the control as unloaded and raising the
+        /// <see cref="Unloaded"/> event.
+        /// </summary>
+        internal void OnUnloadedCore()
+        {
+            if (_isLoaded)
+            {
+                _isLoaded = false;
+                OnUnloaded();
+            }
+        }
+
+        /// <summary>
+        /// Invoked just before the <see cref="Loaded"/> event.
+        /// </summary>
+        protected virtual void OnLoaded()
+        {
+            var eventArgs = new RoutedEventArgs(LoadedEvent);
+            eventArgs.Source = null;
+            RaiseEvent(eventArgs);
+        }
+
+        /// <summary>
+        /// Invoked just before the <see cref="Unloaded"/> event.
+        /// </summary>
+        protected virtual void OnUnloaded()
+        {
+            var eventArgs = new RoutedEventArgs(UnloadedEvent);
+            eventArgs.Source = null;
+            RaiseEvent(eventArgs);
+        }
+
         /// <inheritdoc/>
         protected sealed override void OnAttachedToVisualTreeCore(VisualTreeAttachmentEventArgs e)
         {
             base.OnAttachedToVisualTreeCore(e);
 
             InitializeIfNeeded();
+
+            // Note: This code is not multi-threaded ready and assumes to run only on the UIThread
+            if (_isLoaded == false)
+            {
+                _loadedQueue.Add(this);
+
+                if (_isLoadedProcessingRequested == false)
+                {
+                    _isLoadedProcessingRequested = true;
+
+                    Dispatcher.UIThread.Post(() =>
+                    {
+                        foreach (var item in _loadedQueue)
```

Comment:
There is a potential for "collection changed during enumeration" exception here, since it's valid to add new controls from a `Loaded` callback.

Reply:
Will just add a lock. Let me know if you have a better idea for this.

Reply:
I guess the other option is just to copy the enumeration to a list using linq or a helper method. That may be slower than a lock though, really don't know.

Reply:
Discussed below more: Lock can't be used in this scenario.

Some sort of copy will be required and a better idea is to swap out the hash set for a new one. This is creating hash sets repeatedly though which has overhead. I'll keep going down this path but this is creating a lot of complexity for what may turn out not to be any real performance benefit.

Reply:
Marking resolved. Further discussed below but two hash sets will be used. While one is being processed, the other accepts adding new controls to process next.